salt/_pillar/metalk8s_nodes.py

```diff
@@ -37,7 +38,37 @@ def node_info(node, ca_minion):
             if role:
                 roles.add(role)
 
-    if node["metadata"]["name"] == ca_minion:
+    # Introduced as a workaround in cases which bootstrap setup fails after the
+    # kubernetes API server is up and running but node role labels are still
+    # not set.
+    #
+    # The issue is that when the API server is running the ext_pillar will
+    # overwrite the hardcoded roles with the ones from the API. It will cause
+    #  bootstrap setup failure as described here:
+    # https://github.com/scality/metalk8s/issues/2137
+    #
+    # The `is_bootstrap` flag is set in pillar only when run the
+    # `metalk8s.orchestrate.bootstrap` state.
+    def get_roles_from_pillar(pillar, node_name):
+        if "metalk8s" in pillar:
+            if "nodes" in pillar["metalk8s"]:
+                if node_name in pillar["metalk8s"]["nodes"]:
+                    return pillar["metalk8s"]["nodes"][node_name].get("roles", [])
+
+        return []
+
+    if not roles and pillar.get("is_bootstrap", False):
+        if "metalk8s" in pillar:
+            for role in get_roles_from_pillar(pillar, node_name):
+                log.info(
+                    "Workaround: Adding role '%s' to pillar in node '%s'",
+                    role,
+                    node_name,
+                )
+                roles.add(role)
```

Comment:
To me you do not need for loop here

```suggestion
            roles = set(get_roles_from_pillar(pillar, node_name))
            log.info(
                "Workaround: Adding roles '%s' to pillar in node '%s'",
                roles,
                node_name,
            )
```

Reply:
Done !!!